Allow MNP calling in GVCF mode with stern warnings about not trying …

tfenne opened the pull request:
…joint-genotyping from the resulting GVCFs.

@ldgauthier & @davidbenjamin this PR is a follow up from our short conversation in #4650 a couple of months ago, where I was wanting to generate GVCFs with MNP support.

My goal here is that I really want to be able to generate a single VCF that a) gives me reference confidence and b) gives me MNPs for close by variants.  This is for a clinical pipeline where all calling is done one sample at a time, so the problem of joint-genotyping from different MNP representations doesn't come up.  I did briefly look at using `--emit-ref-confidence BP_RESOLUTION` but that has two issues that make me prefer this route:

1) The generated files are really very large because they have a row for every single BP
2) More problematic, is that when there is a MNP of say `ACG/GCT` two things happen that are less than ideal from my perspective.  The first is that rows are emitted into the VCF for all three positions (the variant at A's position, and two `<NON_REF>` lines at the positions for the C and T respectively).  Secondly, when one or more bases is the same in both MNP alleles (the C in this case) that base is output with a very high hom-ref GQ, which feels wrong!

I'm more than happy to modify this PR to address any concerns you have (e.g. adding a `--force-mnps-with-gvcfs` parameter that has to be specified, or requiring `--unsafe` to enable this).  I'm also open to other solutions, but this seemed expedient and reasonable for folks running single-sample pipelines like you see in clinical settings.

ldgauthier commented:
I'd feel much more comfortable if there was a check in GenotypeGVCFs.  Can you add something to look for MNPs in the ReferenceConfidenceVariantContextMerger.merge() loop over the input vcs and error out if it finds one (and there's more than one input)?

tfenne commented:
@ldgauthier Absolutely.  That sounds like a sensible precaution.  I'll try and make that addition today.

tfenne commented:
@ldgauthier I've added a check in `CombineGVCFs`.  After staring at the code for `GenotypeGVCFs` for a while and scratching my head, I then read the documentation(!) which states:

>  The GATK4 GenotypeGVCFs tool can take only one input track.  Options are 1) a single single-sample GVCF 2) a single multi-sample GVCF created by CombineGVCFs or 3) a GenomicsDB workspace created by GenomicsDBImport.  A sample-level GVCF is produced by HaplotypeCaller with the `-ERC GVCF` setting.

This makes me think that the correct place for this check is actually in `CombineGVCFs` as it should be fine to genotype a single-sample GVCF with MNPs in it, but we want to prevent the creation of multi-sample GVCFs with MNPs.  Does that sound right?

tfenne commented:
No worries.  I'm happy to ada a test or two, just wanted to make sure I was on the right track first.

Re: GenomcisDBImport - I'm sort of hoping that anyone advanced enough to be using GenomicsDB will understand/heed the warnings about not enabling MNPs during GVCF generation, since I'm assuming it's mostly Broad and other large-scale facilities using that?

droazen commented:
@tfenne Well, `GenomicsDBImport` is starting to replace `CombineGVCFs` for some of our userbase as well, so I'm not sure we can assume users of that tool are necessarily more advanced than average...

It should be possible to patch `GenomicsDBImport` to throw a `UserException` whenever it encounters a MNP, if @ldgauthier agrees that that's the right thing to do here. That tool creates a map of `FeatureReaders` up front for its inputs and passes them to GenomicsDB. We could pretty easily wrap the existing FeatureReaders within a custom FeatureReader implementation that throws whenever it sees a MNP (or returns an iterator of VariantContexts that throws).

tfenne commented:
Fair enough @droazen - I'd figured they'd probably be the more advanced users who would be more careful, but I can take a look at adding a check there also.

tfenne commented:
@ldgauthier & @droazen I've done as you've suggested.  There is now a check in `GenomicsDBImport`, by wrapping the FeatureReader.  It's a little ugly but it gets the job done.  I've also added a simple test for GenotypeGVCFs to genotype a GVCF that has an MNP in it.  I _think_ this is probably now ready for review.  Let me know if you think further tests are needed!

ldgauthier commented:
Just the word "subset" gives me a bad feeling in the pit of my stomach because there are been so many bugs with respect to genotypes in subset variants, so that wouldn't be my first choice. In terms of HTSJDK, perhaps the question is how likely is it that the "MNP if you don't count NON_REF" functionality is to occur again?  Honestly it seems not very likely.  I don't know what the timeline for your use case is, but I usually lean towards the most expedient solution anyway.  The first option of keeping the code localized to GATK is entirely fine with me.

tfenne commented:
Thanks @ldgauthier.  I've gone ahead and added a utility method to GATKVariantContextUtils, and tests for that, and used that in CombineGVCFs and GenomicsDBImport.  Those tests pass for me now.  Fingers crossed the travis build(s) pass.

ldgauthier commented:
New utility function looks good to me.

I'm seeing ` java.nio.file.NoSuchFileException: /gatkCloneMountPoint/src/test/resources/org/broadinstitute/hellbender/tools/GenomicsDBImport/mnp.input.g.vcf` in the Travis log.  Looks like you actually put the file in .../tools/CombineGVCFs/ ?

Also there seems to have been something weird that happened to one of the Funcotator files, perhaps during a rebase?

tfenne commented:
Oops, sorry.  The missing file exception was just my forgetting to add files before committing.  I wasn't sure what the right thing to do was - I'm using the sam gVCF for three different tests.  It's tiny, so I added it in three places, so that changing one wouldn't break the other two tests.  Hopefully that was the right thing to do.

I have no idea what happened with that Oncotator file, but I've reset it back to how it was in master, and also squashed my commits, and it seems to have fixed the problem.

ldgauthier commented:
Tests are still failing, but other than that I'm on board.

tfenne commented:
Doh, I see the problem.  There's a test for HC that MNP+GVCF should throw an exception.  Which of course it now doesn't.  I've removed that test and it looks like that was the only one failing.

@droazen do you also want to review?  Or is this ready for merge once the tests pass?

droazen commented:
@tfenne I'll have a (brief) glance at the `GenomicsDBImport` changes this afternoon.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -1846,5 +1846,32 @@ public static boolean isXnp(final String reference, final String alternate) {
+     */
+    public static boolean isMnpWithoutNonRef(final VariantContext ctx) {
+        final List<Allele> alleles = ctx.getAlleles();
+        int length = -1;
```
This code could be more concise as follows:

```
final int refAlleleLength = allleles.get(0).length();
if (alleles.stream().anyMatch(a -> a.length() < 2 || a.length() != refAlleleLength)) {
   return false;
} else {
   return alleles.stream().noneMatch(a -> a.isSymbolic() && a.equals(Allele.NON_REF_ALLELE));
}
```
Usually the decision to `return false` will be made on the first pass, so that's not really a performance issue, although the streams will be a lot slower than the `for` loop.  You can decide if that's a problem.

tfenne replied:
I'd forgotten that the ref allele cannot be symbolic, thanks.  I've used that to make the method more concise, but still using the forloop:

```java
    public static boolean isUnmixedMnpIgnoringNonRef(final VariantContext vc) {
        final List<Allele> alleles = vc.getAlleles();
        int length = vc.getReference().length(); // ref allele cannot be symbolic
        if (length < 2) { return false; }

        for (final Allele a : alleles) {
            if (a.isSymbolic() && !Allele.NON_REF_ALLELE.equals(a)) { return false; }
            else if (!a.isSymbolic() && a.length() != length) { return false; }
        }

        return true;
    }
```

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -635,12 +635,54 @@ public Object onTraversalSuccess() {
+                    @Override public Iterator<VariantContext> iterator() { return this; }
+                    @Override public boolean hasNext() { return this.inner.hasNext(); }
+                    @Override
+                    public VariantContext next() {
```
Throw `NoSuchElementException` if `next()` called when `! hasNext()`

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -635,12 +635,54 @@ public Object onTraversalSuccess() {
         final Function<SeekableByteChannel, SeekableByteChannel> cloudIndexWrapper = (cloudIndexPrefetchBuffer > 0 ? is -> SeekableByteChannelPrefetcher.addPrefetcher(cloudIndexPrefetchBuffer, is) : Function.identity());
         try {
-            return AbstractFeatureReader.getFeatureReader(variantURI, null, new VCFCodec(), true, cloudWrapper, cloudIndexWrapper);
+            final FeatureReader<VariantContext> r = AbstractFeatureReader.getFeatureReader(variantURI, null, new VCFCodec(), true, cloudWrapper, cloudIndexWrapper);
```
r -> reader

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerArgumentCollection.java:
```diff
@@ -133,6 +133,7 @@
     @Argument(fullName = MAX_MNP_DISTANCE_LONG_NAME, shortName = MAX_MNP_DISTANCE_SHORT_NAME,
-            doc = "Two or more phased substitutions separated by this distance or less are merged into MNPs.", optional = true)
+            doc = "Two or more phased substitutions separated by this distance or less are merged into MNPs. " +
+            "When used in conjunction GVCF mode, resulting GVCFs cannot be joint-genotyped.", optional = true)
```
Maybe add a "WARNING:" prefix here to make this disclaimer stand out in the help output.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -1846,5 +1846,32 @@ public static boolean isXnp(final String reference, final String alternate) {
+        int length = -1;
+        for (final Allele a : alleles) {
+            if (a.isSymbolic()) {
+                if (!Allele.NON_REF_ALLELE.equals(a)) return false; // Don't allow any non-NON_REF symbolic alleles
```
Add explicit curly braces around all if statement bodies.

droazen replied:
Yes, totally ok to have them on the same line -- I just care that they actually are there :)

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -1846,5 +1846,32 @@ public static boolean isXnp(final String reference, final String alternate) {
+     * @return true if the context represents an unmixed MNP (i.e. all alleles are non-symbolic and of the same
+     *         length > 1), false otherwise
+     */
+    public static boolean isMnpWithoutNonRef(final VariantContext ctx) {
```
This method name is a little confusing...maybe `isUnmixedMnpIgnoringNonRef()` would be a (slightly) better/more precise name?

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -208,6 +211,17 @@ public void testGenomicsDBThreeLargeSamplesSitesOnlyQuery() throws IOException {
         testGenomicsDBImporterWithGenotypes(LOCAL_GVCFS, intervals, COMBINED_SITES_ONLY, b38_reference_20_21, true, true, true);
     }
 
+    @Test(expectedExceptions={UserException.BadInput.class, CompletionException.class}, expectedExceptionsMessageRegExp=".*GenomicsDBImport does not support GVCFs.*")
```
Why do we expect a `CompletionException` here?

tfenne replied:
I didn't really understand if it was a gradle thing or a GATK/GenomicsDB thing but the code is being run in some async setup, and what I'm actually getting back is a CompletionException which contains the BadInput exception.  If you'd rather I can catch the CompletionException in the test and rethrow the cause?

droazen replied:
@tfenne It looks like you've uncovered a problem in the tool itself. In `GenomicsDBImport.traverse()`, we need to add an additional catch block for `CompletionException` around the `importer.executeImport()` call (The `CompletionException` is being caused by a `future.join()` call deep in the GenomicsDB library). Once you do this, you can remove `CompletionException` from the list of `expectedExceptions` in your test. 

In order to ensure that your MNP exception message will actually be shown to the user, I'd also suggest unpacking the `CompletionException` in that catch block in `GenomicsDBImport` if `e.getCause() instanceof UserException`, and rethrowing `e.getCause()` in that case. It's gross, but I think making the error message visible to the user should take priority here. 

This also made me realize that you should have a second integration test that exercises the `getFeatureReadersInParallel()` path in the tool. You can do this by copying your existing test, adding an overload of `testGenomicsDBImporter()` that takes a `threads` argument and passes it through to `writeToGenomicsDB()` as the final arg, and calling it with threads = 2.

tfenne replied:
Done.  I deviated a little from what you suggested:

1. In `GenomicsDBImport` if the cause of the `CompletionException` is a runtime exception I rethrow that, otherwise the original exception.  I figure it's always going to be more informative to see the actual exception where possible.
2. I didn't overload `testGenomicsDBImporter()`, I just added a `threads` parameter and in the other ~3 places passed `1`.
3. In my actual test, I just added a `for` loop that runs the same code twice with threads = [1, 2].

My thoughts being that (2) and (3) keep things a little more compact rather than duplicating or chaining code.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -2416,4 +2416,26 @@ public void testSplitBiallelicsGenotypes(final VariantContext vc, final List<Var
     }
+
+    /** Makes a list of alleles, with the first one being marked as reference. */
+    public List<Allele> makeAlleles(String... ss) {
```
`private static`

tfenne replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -2416,4 +2416,26 @@ public void testSplitBiallelicsGenotypes(final VariantContext vc, final List<Var
+    /** Makes a list of alleles, with the first one being marked as reference. */
+    public List<Allele> makeAlleles(String... ss) {
+        final List<Allele> alleles = new ArrayList<>();
+        for (int i=0; i<ss.length; ++i) alleles.add(Allele.create(ss[i], i== 0));
```
Braces around loop body

tfenne replied:
Done.

droazen commented:
@tfenne I've responded to the two remaining comments (https://github.com/broadinstitute/gatk/pull/5182#discussion_r221039429 and https://github.com/broadinstitute/gatk/pull/5182#discussion_r221054996)

davidbenjamin commented:
I'm satisfied.
